### ISSUE TITLE
Dandelion: Reduce likelihood of a tx reaching the embargo threshold

### ISSUE THREAD

**Issue by lehnberg** [mimblewimble/grin#1982]
It appears that in the current dandelion config, 15% of all transactions can be expected not to fluff prior to the embargo timer. This means that these transactions would then be broadcasted the traditional way from the sender, making it straight forward to determine which IPs they originate from.

15% seems like a high number for this likelihood, and it may make sense to find ways to reduce it.

### Relevant [Gitter thread](https://gitter.im/grin_community/Lobby?at=5bec1ad762866f74739943a7): 
>Antioch Peverell @antiochp 14:53
we see the embargo timer expire pretty frequently
what we do need to do is investigate this appears to expire more often than not
specifically - we see it expire on the originating node (say I send a stem tx, I see my node's embargo timer expire)
>
>lehnberg @lehnberg 14:55
https://github.com/mimblewimble/grin/blob/fb3f967a860e184a87753b00b6281d10e1c2d4cc/pool/src/types.rs#L35
So with
`patience = 10s`,
`embargo = 180s`,
it would mean that there’s a maximum of `180/10=18 hops` that could occur before embargo clocks out and the originating node broadcasts the transaction the traditional way.
Then the question becomes: What’s the probability of hitting 18 stems in a row when `stem=90%`?
Which (I think) can be expressed as
`0.9^18 ~ 0.15`, i.e. 15%
Am I completely off base here?
>
> ...
>
>valdok @valdok 14:59
@lehnberg you seem to be right. If you have a constant "embargo" timer, as well as constant "patience" timer, then it seems that it'll be activated in 15% of the cases. BTW, in the paper of Dandelion++ they tell that the "embargo" timer should be random, otherwise the origin of the transaction will always disclose itself.
>
> lehnberg @lehnberg 14:59
But if what I calculate above is correct, it would mean that you would expect 15% of all transactions to belong to the IP that broadcasts it, and that doesn’t sound right.
Or good, at least.

**Comment by antiochp** [comment]
Related - #1349.

**Comment by antiochp** [comment]
Just refreshing my memory of exactly how this works and putting some notes here for reference - 

* patience timer runs every 10s
* each run we take "fresh" txs and either "stem" them or "fluff" them
* if we "stem" then we need to wait for another cycle of the patience timer before we send it on to our dandelion relay (it is in "stem" phase for one full run of the 10s patience timer)
  * so a stem tx is "fresh" for 0-10s and then "stem" for an additional 10s (i.e. at a given dandelion relay for 10-20s)
* the embargo timer is randomized by adding rand(0, 30) secs onto the configured time each run (so we look for any txs older than embargo + rand(0, 30) secs

**Comment by lehnberg** [comment]
> the embargo timer is randomized by adding rand(0, 30) secs onto the configured time each run (so we look for any txs older than embargo + rand(0, 30) secs

@antiochp where is this in the code?

**Comment by antiochp** [comment]
https://github.com/mimblewimble/grin/blob/dd8b4fa264ad84420e468e99ebbdb84aed73005f/servers/src/grin/dandelion_monitor.rs#L209-L211

**Comment by quentinlesceller** [comment]
Sorry for the delay in this convo.

> If I read the Dandelion++ paper and the BIP right, each node that a transaction passes through in the stem phase will have their own embargo timer set and each of those nodes will monitor for the transaction. If a node's embargo timer expires without it having seen the transaction in question in the mempool, this particular node will independently trigger fluffing of the transaction. In other words it's not necessarily the node that originally sent the transaction that has to fluff it, which is how I originally had understood it.

Yes

> ...is not correct, as it wouldn't necessarily reveal which node the transaction originated from. Right? So this would make hitting the embargo timer to be less of an issue.

It's more likely that the node that fluff the transaction is actually the initial sender of the tx. Though it is possible that because of randomise embargo timer the second or the third node on the stem fluff the tx. (Based on 60 sec+ (1-30) secs)

> 1. To what degree are we adhering to what's in the BIP specification and the Dandelion++ paper? The paper for example talks about directed 4-regular graphs, and one-to-one forwarding rules, where transactions received from the same source are always routed forward in the same direction during a given epoch. Do we do this?

We do not do this. We are based on a simplified version of Dandelion++. Also we only have 1 Dandelion relay instead of multiples in the original version.

> 2. On aggregation in the stem phase. In the section of the grincon0 talk discussing this @quentinlesceller mentions that B would do a flip for both A and G's transactions, and only if both land on to stem would B aggregate the transactions. How so? Why not immediately aggregate all transactions that arrive during patience and from thereafter treat them as a single transaction that gets a single coinflip?

First, A and G can be aggregated iff they are both to be 'stemmed or 'fluffed'. I remember that we specifically choose to do a per tx coin flip. https://github.com/mimblewimble/grin/issues/819 (Before we had a single coin flip)



> 3. Also, what happens with the embargo timers of two individual transactions once they become aggregated into one?

When a node receives 2 stem transactions, it sets an embargo timer for each one. If both are meant to be 'stemmed' after the coin flip) they will be propagate as one big stem transaction to the next Dandelion relay. The Dandelion relay will receive this big transaction and set a single embargo timer on it( since it's only one transaction). The first node will not change the embargo timers of the two transactions.

**Comment by antiochp** [comment]
Patience is 10s by default which configures the dandelion monitor to run every 10s.
So a node receiving a tx will wait 0s-10s before the next run of the dandelion monitor.
We then tag that tx as `ToFluff` or `ToStem`.
Then on the _subsequent_ run of the dandelion monitor we fluff all the `ToFluff` txs and stem all the `ToStem` txs.

So a tx that is to remain in stem phase will stay on that node for `0-10s + 10s` before being forwarded to the next stem relay node.
So on average 15s per hop.
10 hops puts us at 150s.

And we fluff 10% of the time.

The embargo timer is 180s by default (randomized with additional 0-30s).

We randomize the embargo timer on every run. I don't know enough about probabilities but this _maybe_ means we end up using the lower bound of this over enough runs?
If we randomly choose 0s here then the timer is more likely to fire than if we chose 30s - so they are not really independent runs?
We may need to rethink this randomization.


So I think - 
1. The embargo timer randomization is flawed (maybe significantly)
1. The embargo timer is simply too short right now

**Comment by antiochp** [comment]
Chance of at least 1 hop: 90%
Chance of at least 2 hops: 90% * 90% = 81%
Chance of at least 3 hops: 0.9^3 = 72.9%
...
Chance of at least 10 hops: 0.9^10 = 34.87%
Chance of at least 12 hops: 0.9^12 = 28.24%

12 hops at 15s each = 180s which is the low end of the randomized embargo times.
So I think we actually only have around 70% chance of stemming a tx before the embargo timer expires.

**Comment by lehnberg** [comment]
> So I think we actually only have around 70% chance of stemming a tx before the embargo timer expires.

Of _fluffing_ you mean?

**Comment by antiochp** [comment]
@lehnberg thanks fixed.

**Comment by antiochp** [comment]
(I'm terrible at probabilities by the way so this may all be way off...)
But I think originally we were thinking 10% chance off fluff, translates to 10 hops, so the embargo should be longer than 10 hops. And hops are 10s long.
None of this is really accurate...

**Comment by antiochp** [comment]
And I think to get it down to 1% of txs hitting the embargo timer we'd need something along the lines of:

0.9^44 = 0.97%

44 hops...

Which is an embargo timer of 44 * 15s = 11 minutes... 
Which feels completely counter-intuitive to me.

**Comment by sesam** [comment]
We could:

- tie embargo to block height, so we always wait 1-2 blocks before revealing an un-stemmed tx
- stem faster: 0-10 seconds instead of 10-20

I think [dandelion++](https://arxiv.org/pdf/1805.11060.pdf) forwards a stemming transaction to its stemming peer much faster. Quickly scanning the text, I saw an average delay of 0.3s per node that a bitcoin tx would pass through.

Also for reference #1059.

**Comment by antiochp** [comment]
@sesam - just to be clear, our delay is not part of Dandelion (which is purely around anonymity in terms of where the tx came from). We leverage Dandelion with an added delay so we can attempt to aggregate txs. The delay is purely grin specific.

**Comment by sesam** [comment]
thanks :) When I pulled in dandelion into the discussion originally it was ~weeks after the original paper was released. And I've not yet managed to fully read dandelion++ through.

Re. how we can get more aggregation. Potential stemming peers today must be not firewalled, right? I think that was introduced as a minimalistic protection against someone actively connecting to every grin node with multiple connections to try to occupy a majority of the "edges" of grin's "network graph".

- If the number of firewalled nodes is much larger than potential stemming peers, then we might prefer to send to stem peer and have that one wait. That might be true now initially, but hopefully not later on.

- If we have "enough" potential stemming peers, then one node's first-hop stemming peer quite likely will not get multiple txs before the embargo ends. To get more cut-through action it's probably better to have stemming peers quickly forward anything they've got to other stemming peers. So propagating along a longer route of stems, going faster rather than slower and "vacuuming up" any not-yet-fluffed transactions along the way. It seems this second approach is what you've proposed? And it also seems compatible with the findings in d++, as far as I've read up on it. Might still want to ping gfanti once we have completed our own idea of what to do.
